### DISCUSSION TITLE
Discussion on coupled cluster module

### DISCUSSION THREAD

**Discussion by xubwa** [pyscf/pyscf#2657] [Ideas]
I would like to discuss two default behaviors in the current coupled cluster implementation.
1. Handling two electron integrals.
When running a ccsd(t) calculation after a ccsd calculation, since if no eris object is passed to these two methods, a eris object will be generated and integral transformation will be performed. If one has little knowledge about this behavior, then one extra integral transformation is performed, which is indeed a waste. Should something be done such as after a integral transformation is done, the eris object is assigned as a member of the cc class, so that after performing ccsd calculation, the ccsd(t) calculation no longer requires integral transformation.
2. Semicanonicalizaiton
As of now, when a UCCSD object is created from ROHF object, a semicanonicalization is not done, it would not change the result of ccsd calculation, but do have an influence on the ccsd(t) energy, yielding a meaningless number. Should semicanonicalization at certain step becoming a standard behavior?

**Comment by susilehtola** [comment]
Semicanonicalization will also have an effect on the CCSD energy, if the number of frozen core electrons is non-zero.

**Comment by xubwa** [comment]
It has an effect, but it's minor, one behavior can be chosen as default while leaving the other one doable. It's at least better than doing nothing here imo. Any suggestions on which should be a better practice?

**Comment by fishjojo** [comment]
1. Agree. `eris` can be a member of the CCSD class.
2. Agree. Semicanonicalization can be performed before computing the (T) energy as we don't have iterative (T) solver anyway.

**Comment by xubwa** [comment]
@fishjojo It should be performed before the ccsd calculation since it (t) calculation requires a converged t1 and t2 amplitude. Another point also occurred to me that the file behind an eris object is hard coded as a lib.H5TmpFile, making it hard to be modified to save the integral for a rerun, it might be nice to have such a possibility to have the integral saved.

**Comment by fishjojo** [comment]
I think one could transform the integrals and amplitudes to the semicanonical basis after CCSD, although this introduces more work.
I'm not sure if there is a high demand of saving the MO basis integrals. If people do want to save those, we could simply add an argument such as the file name to the `ao2mo` function.

**Comment by ndattani** [comment]
@fishjojo Is it possible to print semi-canonical ROHF integrals in FCIDUMP format before doing a CCSD calculation? MRCC requires the ROHF integrals to be from semicanonical orbitals. Alternatively, MRCC can use UHF integrals, but PySCF can't print them in FCIDUMP format yet (see this: https://mattermodeling.stackexchange.com/q/13172/5 and https://github.com/pyscf/pyscf/issues/358#issuecomment-2162690658).
